Feature/2020 09 14 introduction to api management (#634)
ordina-jworks/ordina-jworks.github.io: _posts/2020-09-14-Introduction-to-API-Management.md

Keeping an overview on all available API's also becomes a struggle these days.
        
Well, instead of inventing the wheel over and over again, I searched for the best solution and found that an API Management tool can solve a lot of those problems and even add extra value within your landscape.
This tool helps you in managing you APIs, setting the right responsibilities on the right level. 
You can add plenty additional functionalities, but most of all this happens in one tool!
That’s one of the reasons it is important for your business to implement such a tool in your organization.
        
                
## Investigation

### API Management or API Gateway?
API management and API gateway are used interchangeably.
But is it now an API manager, API management tool, or API gateway?
Let’s dive into the naming’s.  
An API management tool refers to the application or overall solution of managing APIs.
This mostly indicates that they are talking about the application and its features.
A management tool can include one or multiple API Gateways, depending on the solution needed in the organization.

> An API gateway is actually a middleware that is placed in front of your services an act as a central point to administrate, route and secure your services.

Author: @pvanhees The changes are adapted.
on your question "is middleware the correct terminology here? " 
I understand it's debatable. 
The gateway controls the exchange of information between the client software and the software that manages the corporate data".
It can change the format of the data (for example changing XML's to JSON and visa versa) to enable communication between (legacy) systems.
Those are a few reasons why I think we can call it middleware

### API landscape without an API management tool
Before going into the details, let’s take a look on how most of the current API landscapes are build and how we can improve those with an API management tool.  
Without an API manager, your APIs will be the direct integration point to the business domain data.
Any integration of for example security, needs to be implemented in your API. This sometimes leads to duplicate code and the fact that your API is not only responsible for exposing the Data, but also for implementing other cross cutting functionalities.
	     
<img src="/img/2020-09-14-Introduction-to-API-Management/WithoutApiManagementTool2.jpg" alt="API landscape without an API management tool" width="350" height="200" class="image fit">

### API landscape with an API management tool
This all changes when we implement an API Management tool.

<img src="/img/2020-09-14-Introduction-to-API-Management/WithApiManagementTool2.jpg" alt="API landscape with an API management tool" width="500" height="350" class="image fit" style="vertical-align:middle;margin-left:2%"/>       

> In this situation an API management tool (API Gateway) is deployed between the client and the APIs.

Reviewer: Like in the sense; 
API landscape with an API Gateway

> In this situation an API management tool (API Gateway) is deployed between the client and the APIs.

Reviewer: API Gateway is the correct naming here

I do hear you thinking already “We’re not going to install an additional tool, only for splitting out security rules from the API’s", but that’s not the only thing an API management tool can do.

### Responsibilities of an API Management tool
I listed out the most important responsibilities for you, so that you can understand the power and the actual benefits of an API Management tool:            
     
1.	Authentication 
    * Enables one of the most important parts of security on an API. Validating the person’s identity!  
    There are different kinds of Authentication:   
        * Http Basic Authentication
            * Using this approach, a user agent simply provides a username and password to prove their authentication.  
            This approach does not require cookies, session ID’s, or login pages because it leverages the HTTP header itself. While simple to use, this method of authentication is vulnerable to attacks that could capture the user’s credentials in transit.
            
        * OAuth
            * OAuth 2.0 is the best choice for identifying personal user accounts and granting proper permissions.  

>             With this method, the user logs into a system. That system will then request authentication, usually in the form of a token. The user will then forward this request to an authentication server, which will either reject or allow this authentication.

Author: Hi Pieter,
I understand what you mean with "Logs in in a system" but it doesn't seem correct.
I changed it to "With this method the user logs in to a system."

2.	Authorization 
    * Enables fine grained authorization to API resources based on authenticated user roles.  
    This involves checking resources that the user is authorized to access or modify via defined roles.
3.  Caching resources
    * Allows caching of API responses in the Gateway to reduce overall traffic to the back-end API.  
    Latency will be improved because the API gateway can talk directly with the backend service or can use the cached resource and doesn't have to fetch the information multiple times.  
    With caching you also avoid that the backend service will be overloaded by requesting the same information repeatedly.  
    Resources can be cashed for a specified time-to-live (TTL) period.

>     The gateway will retrieve the resource again from the backend service once the time has passed, and a new request comes in.

Reviewer: remove ' and a new request comes in'

4.  White/Blacklisting possibilities
    * Allow/Block calls to specific APIs.
    * Allow/Block all calls from a given application.
    * Allow/Block requests coming from a specific IP address.
    * Allow/Block a specific user from accessing APIs.
5.  Time Restricted Access
    * Requests matching the regular expression but made outside the specified time period will receive an error code.  
    This is used to allow access to an API only during certain times.    
6.  Quota’s 
    * Rate limiting

>     	* Enforces rate configurable request limits on an API.  

Author: I changed it to "Enables to set a rate-configurable limit to requests on an API"

8.  Transformation
    * Transformation enables to convert an API format between for example from JSON and XML. 
    If an API is implemented to return XML, but a client would prefer to receive JSON data, this policy can be used to automatically convert both the request and response bodies. In this way, the client can work with JSON data even though the back-end API requires XML.
9.  Monitoring
    * Some tools let you visualize, query, route, archive, and take actions on the metrics or logs like for example the Azure API management tool.  
    Others just foresee a basic around metrics or include a third-party tool like Elasticsearch to do the job.
    
A lot more can be added or customized, depending on the tool of course. There is a wide range of products available, so please refer to the product list below if you want to know the ones used currently.  
To get hands-on experience, I mostly focused on the Open source tools like Apiman, Kong API Gateway, WSO2 API Manager.

### Challenges

There are a few things to consider before you select the right tool for your organization.  

> One of the most important thing is think about the high availability of the services that you provide.  

Author: Hi Pieter,
I like your proposal, but high availability is wildly used, and that's also why I used this in this blogpost.
I prefer to keep it, or is it really bothering you?

In this example you can see that 2 different Gateways are created to separate mobile from other request. The endpoint of the mobile gateway will in this case be different from the Web gateway.
If for any reason at all the Mobile API Gateway gets stuck, the client will still be able to connect to the other Gateway and request for the data he/she needs.

#### Clustering gateways
Another way to ensure high availability is for example to cluster your API Gateways.
Combined with a scalable microservice architecture, a high availability API gateway cluster ensures your application can handle large volumes of traffic and react to unexpected spikes, while being resilient to hardware failures.  
There are big differences between products when you want to cluster an API gateway.
Most products do recommend or even require to setup a Load balancer in front of the API gateway cluster. In fact, to avoid another single point of failure, it is recommended to be able to scale this load balancer as well. Or have a failover scenario in place in case the load balancer crashes. This way you can be sure that your services are always available.  

> Some other products require to scale additional tools like for example Elasticsearch.  

Reviewer 2: ```suggestion
Some other products require you to scale additional tools like for example Elasticsearch.  

```

> An advantage of this solution is that the client always points to the same endpoint.

Reviewer 2: ```suggestion
An advantage of this solution with a load balancer, is that the client always connects to the same endpoint.

```

## API management products

I referred to the Gartner magic quadrant to get the list of most common API management tools currently on the market.
You can find most of them here in alphabetical order.

<img src="/img/2020-09-14-Introduction-to-API-Management/GartnerMagicQuadrant.JPG" alt="Gartner Magic Quadrant" width="400" height="400" class="image fit" style="float:right    ;horzontal-align:middle;margin-left:2%;margin-right:50%;vertical-align:right" />

* Red Hat 3 Scale 
* Akamai API Gateway  
* Akana API Management
* API Man (open source)
* Apigee API Management
* AWS API Gateway
* Axway - AMPLIFY API Management
* Azure API Gateway
* Boomi
* Broadcom
* CA API Gateway
* Express API Gateway (open source)
* Fusio API Management (open source)
* IBM API Connect
* Kong API Gateway (open source)
* Loopback API Framework (open source)
* MuleSoft Anypoint API Management
* Oracle API Manager
* SAP API Manager
* Sensedia API Management Platform
* Sentinet 
* Software AG API Gateway
* TIBCO Mashery
* Tyk API Gateway (open source)
* WSO2 API Manager (open source)

## Hands-on
### Installation

> Just to try a few things, I installed Apiman (and the required server) as a standalone solution on to my laptop and used one of my earlier created API's to test.

Reviewer 2: ```suggestion
Just to try a few things, We installed Apiman (and the required server) as a standalone solution on a laptop and used some existing APIs to test.

```

> It was actually quite simple, however I did download a few versions of it, as some didn't work from the 1st minute.  

Reviewer 2: ```suggestion
It was actually quite simple, however we did download a few versions of it, as some didn't work immediately.  

```

One of the main advantages is that it has some pre-configured settings which made it easy to start.  
Apiman is an Open Source tool, so a lot of documentation and even source code examples are available on the web.  
If you prefer, it is also possible to dockerize it and run this in a container.  

> The latest version of Apiman can be found <a href="http://www.apiman.io" target="_blank" rel="noopener noreferrer">here</a>, or in case you want to try earlier versions or embedded in another server, you can also look in <a href="https://apiman.gitbooks.io/apiman-installation-guide/content/installation-guide/servlet/install.html#_installing_in_jboss_eap_7" target="_blank" rel="noopener noreferrer">this</a> link.  

Reviewer 2: ```suggestion
The latest version of Apiman can be found <a href="http://www.apiman.io" target="_blank" rel="noopener noreferrer">here</a>, or in case you want to try earlier versions or want to try Apiman embedded in another server, you can also look in <a href="https://apiman.gitbooks.io/apiman-installation-guide/content/installation-guide/servlet/install.html#_installing_in_jboss_eap_7" target="_blank" rel="noopener noreferrer">this</a> link.  

```

> Once installed, I started to play around with it.  

Reviewer 2: ```suggestion
Once installed, we started to play around with it.  

```

<img src="/img/2020-09-14-Introduction-to-API-Management/APIMAN_Login.PNG" alt="APIMAN Login screen" width="600" height="375" class="image fit" style="vertical-align:middle;margin-left:2%" />

### Setup

> I did some research on how to set it up, and on the key aspects of Apiman.  

Reviewer 2: ```suggestion
We did some research on how to set it up, and on the key aspects of Apiman.  

```

> It's not my intention to go to much in dept of Apiman or the settings, but a few base principles I want to highlight to help you understand how easy it was to try it out and play with it.  

Reviewer 2: ```suggestion
It's not our intention to go too much in depth of Apiman or the settings, but we want to highlight a few base principles to help you understand how easy it was to try it out and play with it.  

```

The first very important part in Apiman was that everything starts with an Organization.  

> 2ndly you need to setup a plan, which is basically a collection of policies that will be applied to requests made to Services being access through it.  

Reviewer 2: ```suggestion
Secondly you need to set up a plan, which is basically a collection of policies that will be applied to the incoming requests.  

```

> As last you need to setup a Service. 

Reviewer 2: ```suggestion
Finally you need to setup a service. 

```

A service contract is simply a link between an application and a service through a plan offered by that service. 
When a service contract is created, the system generates a unique API key specific to that contract. 
All requests made to the service through the API Gateway must include this API key.  

Apiman start Screen.  
<img src="/img/2020-09-14-Introduction-to-API-Management/APIMAN_Indexpage.PNG" alt="APIMAN start screen" width="600" height="375" class="image fit" style="vertical-align:middle;margin-left:2%" />

Setting up an organization (OrgHome) and creating services.

<img src="/img/2020-09-14-Introduction-to-API-Management/APIMAN_Organization-API.PNG" alt="APIMAN Organization and Services" width="800" height="300" class="image fit" style="vertical-align:middle;margin-left:2%" />

In this example you can see a Customer API which fetches Customer information from a Spring Boot application running in IntelliJ.  

> Every service can be marked public, so that everyone can access it, or you can for example secure it by creating a Client App and generate a required key via a plan.   

Reviewer 2: ```suggestion
Every service can be marked public, so that everyone can access it, or you can for example secure it by creating a client app and generate a required key via a plan.   

```

> In next screen you can see that I created a 'SearchCustomers' client app, where the client needs to use the API-Key in the request to get the response back from our service.

Reviewer 2: ```suggestion
In the screenshot below you can see that we created a 'SearchCustomers' client app, where the client needs to use the API-Key in the request to get the response back from our service.

```

Our service (Spring Boot application) didn't implement any security or authority for this.
Apiman will take the responsibility of implementing this.

> Only if the client uses the right endpoint with the correct API-key, Apiman will retrieve the requested data from our backend service and send the response back to the client.

Reviewer 2: ```suggestion
Only if the client uses the right endpoint with the correct API-key, Apiman will retrieve the requested data from our backend service and sends the response back to the client.

```

 
<img src="/img/2020-09-14-Introduction-to-API-Management/APIMAN_ClientApps-APIKey.PNG" alt="APIMAN Client apps and API-Key" width="800" height="400" class="image fit" style="vertical-align:middle;margin-left:2%" />

### Automation

> As last I wanted to be sure that we can use the tool in an automated way, so I tried to use the Apiman REST API to create new organizations, plans and services.

Reviewer 2: ```suggestion
Finally we wanted to be sure that we can use the tool in an automated way, so I tried to use the Apiman REST API to create new organizations, plans and services.

```

> I had some struggles to make it work, but eventually I was able to create everything I wanted.

Reviewer 2: ```suggestion
We had some struggles to make it work, but eventually we were able to create everything we wanted.

```

Via the REST API we have the possibility to check, add or remove anything you want.

> The build in Apiman-UI is also build on these REST API, so anything you can do in the application should be possible to execute via the API service.

Reviewer 2: ```suggestion
The built in Apiman-UI is also built on these REST APIs, so anything you can do in the application should be possible to execute via the API service.

```

> More on this can be find in the <a href="http://www.apiman.io/latest/api-manager-restdocs.html" target="_blank" rel="noopener noreferrer">API-manager-restdocs</a>.

Reviewer 2: ```suggestion
More on this can be found in the <a href="http://www.apiman.io/latest/api-manager-restdocs.html" target="_blank" rel="noopener noreferrer">API-manager-restdocs</a>.

```

> One sidenote, the downloaded tomcat server does not come with integrated Keycloak server, so I was not able to add users via the REST API. If you're running the standalone WildFly server, this would be integrated by default.  

Reviewer 2: ```suggestion
One sidenote, the downloaded tomcat server does not come with an integrated Keycloak server, so we were not able to add users via the REST API. If you're running the standalone WildFly server, this would be integrated by default.  

```

> I do was thinking to create a user database and let Apiman connect and check for the right credentials and user type.
> But eventually to save some time, I added the users directly in Apiman to test. This is not the preferred way in a production like environment, but it did the job for now.

Reviewer 2: ```suggestion

```
This provides no additional information to the reader, I would leave it out

       
## Conclusion
There are multiple reasons why you should consider implementing an API Management tool in your organization.

> Think about the benefits and segregation of duties/responsibilities of your API’s and the API management tool.  

Reviewer 2: ```suggestion
Think about the benefits and segregation of duties/responsibilities between your APIs and the API management tool.  

```
Is my suggestion correct? Or did I misinterpret what you were trying to say?

You will be able to manage and control your APIs in one tool.
You can implement and adjust security levels like Authentication and authorization in one place.  
This tool makes it easy to add additional policies to different backend services. It can improve current latency problems and will improve the maintainability and scalability of your API landscape.  
If you are still not convinced, take the time to try out a few. 

> For example, Apiman has some prebuild versions that you can download and use immediately.  

Reviewer 2: ```suggestion
For example, Apiman has some prebuilt versions that you can download and use immediately.  

```

> Even if you only have a few API’s running in your organization, you probably will benefit from using such a tool to manage your Security, Routing, or other Policies you want to implement.  

Reviewer 2: ```suggestion
Even if you only have a few APIs running in your organization, you probably will benefit from using such a tool to manage your security, routing, or other policies you want to implement.  

```

> With an API management tool you can make sure that all your APIs just do what they need to do (sharing business related data), and that the API management tool takes care of the other cross cutting functionalities where it is required.  

Reviewer 2: ```suggestion
With an API management tool you can make sure that all your APIs just do what they need to do (sharing business related data), and that the API management tool takes care of the other cross-cutting functionalities where it is required.  

```

image: /img/kris_Jordens.jpg
tags: [API Management, API Gateway, Products, Responsibilities]
category: API
comments: true
---

        
# Introduction to API Management

## Table of contents
* [Introduction to API Management](#Introduction-to-API-Management)
    * [Problem description](#Problem-description)
    * [Investigation](#Investigation)
        * [API Management or API Gateway?](#API-Management-or-API-Gateway)
        * [API landscape without an API management tool](#API-landscape-without-an-API-management-tool)
        * [API landscape with an API management tool](#API-landscape-with-an-API-management-tool)
        * [Responsibilities of an API Management tool](#Responsibilities-of-an-API-Management-tool)
        * [Challenges](#Challenges)
            * [Multiple API gateways configured](#Multiple-API-gateways-configured)
            * [Clustering gateways](#Clustering-gateways)            
    * [API management products](#API-management-products) 
    * [Hands-on](#Hands-on)
        * [Installation](#Installation)
        * [Setup](#Setup)
        * [Automation](#Automation)
    * [Conclusion](#Conclusion)

## Problem description

> In our current digital world where everyone is creating APIs for exposing data internally and for external partners, we should look on how we can manage and control them.

Author: Hi Kevin,
I prefer to keep the part of "exposing data internally and for external partners".
Do you really want to remove it? Or can we maybe find another way to write it down?

> In our current digital world where everyone is creating APIs for exposing data internally and for external partners, we should look on how we can manage and control them.

Reviewer: Maybe in the form of monetizing APIs for external partners or use it for internal usage

> In our current digital world where everyone is creating APIs for exposing data internally and for external partners, we should look on how we can manage and control them.

Author: I changed it to "In our current digital world where everyone is creating APIs to expose data for internal usage, or monetizing APIs for external partners,  we should look at how we can manage and control them."
Let me know what you think about it .

Most of us also know that we not only have to build the API, but we need to foresee additional responsibilities with it. Some that are just a necessary evil. Like for example Authorization, Authentication, or certain routing policies, etc.  

> Keeping an overview on all available API's can become a struggle these days.

Author: The first sentence doesn't seem right (most of us ...)
do you think it may be better that we change it to "Most of us know that in addition to building the API, we also need to provide additional responsibilities."  ?

> Keeping an overview on all available API's can become a struggle these days.

Reviewer: super 👍

        
Well, instead of trying to reinvent the wheel, I searched for the best solution and found that an API Management tool can solve a lot of those problems and add extra value within your landscape.

An API management tool helps you in managing your APIs, setting the right responsibilities on the right level. 

> You can add plenty additional functionalities, but most of all this happens in one tool!

Author: Maybe another suggestion , what do you think about :
"You can add a lot of extra functionality and most importantly it can be implemented in one tool!" ?

> You can add plenty additional functionalities, but most of all this happens in one tool!

Reviewer: Perfect, I like the way you're thinking, nice addition

        * [Setup](#Setup)
        * [Automation](#Automation)
    * [Conclusion](#Conclusion)

## Problem description

In our current digital world where everyone is creating APIs, we should look at how we can manage and control them.

Most of us know that we not only have to build the API but foresee additional responsibilities. 
Some are just a necessary evil, like adding cross-cutting concerns.  
Keeping an overview of all available API's can become a struggle these days.

        
Well, instead of trying to reinvent the wheel, I searched for the best solution and found that an API Management tool can solve a lot of those problems and add extra value within your landscape.

An API management tool helps you in managing your APIs, setting the responsibilities on the right level. 

You can add plenty additional functionalities, but most of all this happens in one tool!
That’s one of the reasons it is important for your business to implement such a tool in your organization.
        
                
## Investigation

### API Management or API Gateway?
The terms API management and API gateway are often used interchangeably, which is not correct.

> An API management tool refers to the application or overall solution of managing APIs.

Reviewer: API Gateway is the main component of API management solutions.
an API management solution consists of more components: API Gateway, API portal, API Governance

> An API management tool refers to the application or overall solution of managing APIs.

Author: Hi Kevin,
What do you mean, or what change would you suggest with this comment?

I do think that it can but doesn't have to be a SAASs solution. 
Certain API Management applications can be hosted and managed 'on premise' by the customer.

> An API management tool refers to the application or overall solution of managing APIs.

Reviewer: look at the image of this site: https://www.softwaretestinghelp.com/api-management-tools/
I think you want to mean that an API Management tool is a set of components, including the API gateway no which can be deployed on-premise or use of SAAS solutions

> An API management tool refers to the application or overall solution of managing APIs.

Author: That's exactly what I meant .
I rephrased this section a bit.
I didn't include the on-premise or Saas solution part, as this section is actually about the naming's.
can you let me know if it's more clear , or if I you have some other suggestions?

### API landscape without an API management tool
Before going into the details, let’s take a look on how most of the current API landscapes are build and how we can improve those with an API management tool.  
Without an API manager, your APIs will be the direct integration point to the business domain data.
Any integration of for example security, needs to be implemented in your API. This sometimes leads to duplicate code and the fact that your API is not only responsible for exposing the Data, but also for implementing other cross cutting functionalities.
	     
<img src="/img/2020-09-14-Introduction-to-API-Management/WithoutApiManagementTool2.jpg" alt="API landscape without an API management tool" width="350" height="200" class="image fit">

### API landscape with an API management tool
This all changes when we implement an API Management tool.

<img src="/img/2020-09-14-Introduction-to-API-Management/WithApiManagementTool2.jpg" alt="API landscape with an API management tool" width="500" height="350" class="image fit" style="vertical-align:middle;margin-left:2%"/>       

In this situation an API management tool (API Gateway) is deployed between the client and the APIs.

> All request first have to pass the API management tool, before they are forwarded to the right service.

Reviewer: ```suggestion
As a single entry point, all requests have to pass the API gateway before they are forwarded to the right service.

```

> If the Gateway is installed in the same organizational structure, it’s not required to have your API’s implementing additional cross cutting functionalities.

Reviewer: Debatable, some services might need stricter constraints, the gateway can set a baseline here. 
Cross-cutting concerns are very broad so it's maybe better to rephrase this

Author: It was a proposal from Ken to use Cross cutting func. 
Initially I used a small summary of the most important parts.
You will see that it's coming back multiple times in the blog post.
I didn't find a better way to rephrase it. do you maybe have an suggestion?

Reviewer: Maybe in the form of: it will be possible to move most of the cross-cutting concerns to the API Gateway

Author: Let me know what you think about the rephrased section (after I pushed it :) )

            * Using this approach, a user agent simply provides a username and password to prove their authentication.  
            This approach does not require cookies, session ID’s, or login pages because it leverages the HTTP header itself. While simple to use, this method of authentication is vulnerable to attacks that could capture the user’s credentials in transit.
            
        * OAuth
            * OAuth 2.0 is the best choice for identifying personal user accounts and granting proper permissions.  
            With this method, the user logs into a system. That system will then request authentication, usually in the form of a token. The user will then forward this request to an authentication server, which will either reject or allow this authentication.
        
        * OpenID Connect & OAuth 2.0 API
            * OpenID Connect is a simple identity layer on top of the OAuth 2.0 protocol, which allows computing clients to verify the identity of an end-user based on the authentication performed by an authorization server, as well as to obtain basic profile information about the end-user in an interoperable and REST-like manner.
        
        * API Keys

>             * An API key is an identifier meant to identify the origin of web service requests (or similar types of requests).  

Reviewer: It's just a way to identify the user's identity, not the origin of web service requests

Author: oops, I Copied it from the user guide source

            A key is generated the first time a user attempts to gain authorized access to a system through registration. From there, the API key becomes associated with a secret token, and is submitted alongside requests going forward. When the user attempts to re-enter the system, their unique key is used to prove that they are the same user as before. This API Authentication Method is very fast and reliable but is frequently misused. More importantly, this method of authentication is not a method of authorization.   
2.	Authorization 
    * Enables fine grained authorization to API resources based on authenticated user roles.  

>     This involves checking resources that the user is authorized to access or modify via defined roles.

Reviewer: The API gateway just verifies the token with an external authorization service and propagates it to the downstream service. Here the service will enable fine-grained authorization to API resources.

Author: I changed it

3.  Caching resources

>     * Allows caching of API responses in the Gateway to reduce overall traffic to the back-end API.  

Reviewer: Put a warning that this must be carefully considered
